test/benchmarks/benchmark_slidehash.cc

```diff
@@ -0,0 +1,78 @@
+#include <stdint.h>
+#include <stdint.h>
+#include <limits.h>
+
+#include "zbuild.h"
+
+extern "C" {
+#include "deflate.h"
+}
+
+#include "cpu_features.h"
+
+#include <benchmark/benchmark.h>
+
+#define MAX_RANDOM_INTS 32768
+
+class slide_hash: public benchmark::Fixture {
+public:
+    uint16_t *l0;
+    uint16_t *l1;
+    deflate_state *s_g;
+    void SetUp(const ::benchmark::State& state) {
+        int32_t random_ints_size = MAX_RANDOM_INTS * sizeof(uint16_t);
+        l0 = (uint16_t *)malloc(HASH_SIZE * sizeof(uint16_t));
+        l1 = (uint16_t *)malloc(random_ints_size);
+
+        for (int32_t i = 0; i < MAX_RANDOM_INTS; i++) {
+            l1[i] = rand();
+        }
+
+        for (int32_t i = 0; i < HASH_SIZE; i++) {
+            l0[i] = rand();
+        }
+
+        deflate_state *s = (deflate_state*)malloc(sizeof(deflate_state));
+        s->head = l0;
+        s->prev = l1;
+        s_g = s;
+    }
+
+    void TearDown(const ::benchmark::State& state) {
+        free(l0);
+        free(l1);
+    }
+};
+
+BENCHMARK_DEFINE_F(slide_hash, c)(benchmark::State& state) {
```

Comment:
Hmm, the fixtures don't seem to be buying us a whole lot.  Don't suppose there's a good way to condense this into a preprocessor or constexpr statement where the function pointer gets subbed in and a cpu feature support flag is checked prior to adding it to the benchmark list?  It'd be great if we could do say:
`SLIDEHASH_BENCHMARK("slide_hash_avx2", slide_hash_avx2, x86_has_avx2);`
Instead of having to copy and paste that entire function body + function declaration per implementation.

Actually I feel like the most obvious thing to do is to write one test and parameterize out the function pointer.  Then we just declare variants by switching out the function pointer.  That should be simple, no?

Reply:
Using the fixtures isolates the source code for each test which I like. But there might be some way to make a single function as you say. But in recent commit, I also have a check that checks that the CPU supports the instruction set, which might be harder to parameterize.

Reply:
I'm not familiar enough with constexpr to know about that.

Reply:
I mean it's better, that's for sure.  It's still about 6 lines for something I'd have hoped would be 1 or 2.  But, 2 of those lines are compile time checks for the support.

Reply:
Actually what you have there now I think probably could be wrapped into a macro since they all have the same exact pattern, but you'd still need the compile time checks surrounding the macro.

Reply:
Google benchmark supports Setup/Teardown without having to restore to fixtures....
https://github.com/google/benchmark/blob/main/docs/user_guide.md#setupteardown
So I wouldn't need to use them..